raiden/raiden_event_handler.py

```diff
@@ -357,7 +357,16 @@ def handle_contract_send_channelunlock(
             )
             state_change_identifier = event_record.state_change_identifier
         else:
-            state_change_identifier = 0
+            # In the case that someone else sent the unlock we do nothing
```

Comment:
@konradkonrad In the case of `is_our_unlock` or `is_partner_unlock` being `False` means that the locksroots values in the smart contract are empty. Therefore, i think the implementation here is good.
In the case the returned locksroots are not empty, then we query the database. If we don't manage to find the state which contains the balance hash we're looking for (by `get_state_change_with_balance_proof` or `get_event_with_balance_proof`), that's when the state_change_identifier will be 0 and we should be raising the exception.

Reply:
I agree with @konradkonrad here, this logic will hide errors. @rakanalh 's argument is correct *for the problem stated in the issue*, however, if the call to update transfer failed to update the locksroot, this will hide the fact the unlock never happened.

Additionally, the call to unlock should never have happened, because the on-chain unlock is handled [here](https://github.com/raiden-network/raiden/blob/master/raiden/transfer/channel.py#L1859-L1861). (well, there is still the chance of a race without #2885)

Reply:
> Additionally, the call to unlock should never have happened, because the on-chain unlock is handled here. (well, there is still the chance of a race without #2885)

Someone else could have called unlock. Or even ourselves before restart. The previous logic caused multiple raiden node crashes due to races like that during the December testing.

> however, if the call to update transfer failed to update the locksroot, this will hide the fact the unlock never happened.

There is a warning in there now. Want something stronger?

Reply:
> Someone else could have called unlock. Or even ourselves before restart

I understand that, what does that have to do with the linked code?

> There is a warning in there now. Want something stronger?

Unless we change the recoverable/unrecoverable definitions, yes.

Reply:
> Unless we change the recoverable/unrecoverable definitions, yes.

Yes I agree, if indeed this can happen (neither `partner_unlock` or `our_unlock` for other valid reasons, then this may hide errors and we should change the logic.

So let me see if I am following what you are saying correctly. We may locally expect that a partner has pending locks in his state and [generate](https://github.com/raiden-network/raiden/blob/1466689bd21bf7a4e285784bc502b3620fdc44ba/raiden/transfer/channel.py#L1792) a `ContractSendChannelBatchUnlock` during `handle_channel_settled`.

Let's assume we closed the channel, and he may for whatever reason not have called `updateTransfer()`.

In that case then the onchain locksroot would be an `EMPTY_HASH` and the linked logic would break. Is this what you are speaking about?

In the case where we are sending an onchain unlock for all pending locks from `partner` to `participant` the onchain locksroot **must be** the same as the offchain one. This is enforced by the [contract](https://github.com/raiden-network/raiden-contracts/blob/ea259567c8d7b06410e7969d5a653953eaf39be9/raiden_contracts/contracts/TokenNetwork.sol#L814-L823). So I would say in that case the state machine should never have generated the `ContractSendChannelBatchUnlock`. This does not seem to be the case at the moment:

https://github.com/raiden-network/raiden/blob/1466689bd21bf7a4e285784bc502b3620fdc44ba/raiden/transfer/channel.py#L1788-1807

So perhaps we should add a check inside `handle_channel_settled` to see if the onchain locksroot matches what we expect to send and if not, just don't generate the `ContractSendChannelBatchUnlock`?

Reply:
> In that case then the onchain locksroot would be an EMPTY_HASH and the linked logic would break. Is this what you are speaking about?

Yes, break in the sense that it will hide the hypothetical bug which resulted in `update` not being called. Hiding the root error.